### PR TITLE
Issue/57

### PR DESCRIPTION
**Pull request by hazmat345**
This fixes beer-garden/beer-garden#57.

How we handle output from the plugins is now more explicit. Previously we had one logger in the PluginRunner, now we have three.

Logging generated by the PluginRunner itself will use the 'normal' log formatting - timestamp, level, logger name, message.

Logging that gets pulled out of the plugin subprocess will use one of two loggers:
- If we can't find a level in the message then we will log using the `timestamp` logger. This adds a timestamp to the message but otherwise just passes the message through. Since we have no idea what the 'true' level should be we log these at the `ERROR` level to ensure they get outputted (but that level is *not* part of the log message).
- If we do fine a level then we assume the plugin is using its own logger with its own formatter. We will log the message at the level we found but will not mangle the message in any way.

Also, the formatting no longer changes based on if you're logging to a plugin log file or STDOUT. I think that causes a good amount of confusion (at least it did for me), but it does make it harder to interpret what's happening if you're looking at STDOUT with a bunch of plugins running at the same time. It would be easy to switch back.

I also converted some tests to pytest style.

### CODE REVIEW

**Comment by loganasherjones** [comment]
So, I think what we would like to do is the following:

* Split out `STDOUT` and `STDERR` logging and handle them separately
* `STDOUT` which would include `print` statements will be logged at the `INFO` level. 
* `STDERR` which includes standard `logging` output will need to be searched for the appropriate level, if no level is found, we can assume it should be logged at the `error` level.

I would like to confirm:

* Can you specify in python that logging go to `STDOUT`? I would assume so, if so, we probably still need to search the output in `STDOUT` for log levels.
* Stacktraces get printed to `STDERR`

**Comment by hazmat345** [comment]
We can definitely split them out and handle them distinctly. I agree we should do this.

Points:

If you have a plugin that doesn't do any logging configuration then the 'default' behavior is to have the root logger write to `stderr`. This means that a local plugin that raises in its `__main__` (without configuring logging) is going to be writing the stacktrace to `stderr`. However, if you configure your logger to write everything to `stdout` that's where it'll write the stacktrace.

Also, you can do this (not sure why you'd want to, but you can, it's easy):
```python
print('Hi!', file=sys.stderr)
```

So I think whatever we decide as far as handling message with / without loglevels we should be consistent across `stdout` and `stderr`.

Because of this, I think we should treat them the same except for having different default levels. The plugin logs at whatever level it was configured to log at. Messages, whether coming from `stdout` or `stderr`, are inspected for a level. If we find one, we try to log at that level. If we don't, `stdout` logs at INFO and `stdout` logs at ERROR.

This would mean that local plugins having a default loglevel of WARNING would capture / write stacktraces to the plugin log file, but they would not capture `print` calls. I think that's what we want?

We could maybe add something to the message to indicate the source (`out` vs `err`)?

Answers to confirms:
- It can. That's actually how we define the brewtils `default` hanlder.
- They do if you're using `logging.basicConfig`, but they are just log messages. They'll go where the logger is configured to send them.

File: bartender/local_plugins/plugin_runner.py [hunk 1]
```diff
@@ -135,34 +141,31 @@ def run(self):
...
+        one of the log levels that the python logger knows about we assume that
+        the plugin has its own logger and formatter. In that case we log to our
+        unformatted logger at the same level to keep the original formatting.
 
-        If the plugin is just using STDOUT with no logging then generally the logging level will
-        not be included and so we just add the normal log formatting.
+        If we can't determine the log level then we'll add a timestamp to the
+        start and log the message at the ERROR level. That way we guarantee
+        messages are outputted (this is usually caused by a plugin writing to
+        STDOUT / STDOUT directly or raising an exception with a stacktrace).
```
Comment (loganasherjones) [inline bartender/local_plugins/plugin_runner.py]: Should say `STDOUT / STDERR`

**Comment by hazmat345** [comment]
Sorry, this isn't quite ready yet - still need to make the change to split out the two stream readers.

File: bartender/local_plugins/plugin_runner.py (hunk 1 above)
Comment (hazmat345) [reply bartender/local_plugins/plugin_runner.py]: +1

**Comment by hazmat345** [comment]
Just to be sure, are you OK with this:

This would mean that local plugins having a default loglevel of WARNING would capture / write stacktraces to the plugin log file, but they would not capture print calls. I think that's what we want?

**Comment by hazmat345** [comment]
NEW DEVELOPMENT:

Actually, this is what we currently do. Default log level for local plugins should be INFO, so we'll see `print`s in the plugin logs.

**Comment by hazmat345** [comment]
This is read for another review.

File: bartender/local_plugins/logger.py [hunk 2]
```diff
@@ -20,30 +21,45 @@ def __getattr__(self, attr):
...
     """
     Get a Plugin Logger. If formatted is set to true, do no special formatting.
     Otherwise, make the logger log to something semi-pretty
 
-    :param name:
-    :param formatted:
-    :return:
+    Args:
+        name (str): The name of the logger to create
+        formatted (str / bool): The formatting to apply to the logger
```
Comment (loganasherjones) [inline bartender/local_plugins/logger.py]: Maybe instead of using `formatted` as a `str` or a `bool` we should just pass in the format string itself?

File: requirements.txt [hunk 3]
```diff
@@ -41,7 +42,8 @@ pyflakes==1.6.0           # via flake8
 pygments==2.2.0           # via sphinx
 pymongo==3.5.1            # via mongoengine
 pyrabbit2==1.0.3
-pytest==3.4.2
+pytest==3.5.0
+pytest-mock==1.9.0
```
Comment (loganasherjones) [inline requirements.txt]: Should pytest-mock be added to the requirements.in?

Comment (hazmat345) [reply requirements.txt]: Arg. Good catch.

File: bartender/local_plugins/logger.py (hunk 2 above)
Comment (hazmat345) [reply bartender/local_plugins/logger.py]: Oh good idea

**Comment by hazmat345** [comment]
@loganasherjones this is ready for another review. I also added a test case for a unicode character (I went with _beer mug_ instead of _pile of poo_, don't worry)

**Comment by loganasherjones** [comment]
👍 LGTM
